### PR TITLE
Fix #5210 Don't refer to `index.html` when Haddock will not create it

### LINKED ISSUE #5210: `cabal new-haddock --haddock-for-hackage` claims to have created non-existent files
```
$ cabal get bytestring
$ cd bytestring-0.10.8.2
$ cabal new-configure -w ghc-8.4.1
$ cabal new-haddock --haddock-for-hackage
…
Documentation created:
/home/simon/src/bytestring/dist-newstyle/build/x86_64-linux/ghc-8.4.1/bytestring-0.10.9.0/doc/html/bytestring-0.10.9.0-docs/index.html,
/home/simon/src/bytestring/dist-newstyle/build/x86_64-linux/ghc-8.4.1/bytestring-0.10.9.0/doc/html/bytestring-0.10.9.0-docs/bytestring.txt
Documentation tarball created:
/home/simon/src/bytestring/dist-newstyle/bytestring-0.10.9.0-docs.tar.gz
```

But neither the `index.html` and `bytestring.txt` files nor the tarball exist.

### PR DESCRIPTION
**Pull request by mpilgrem**
Haddock documents that it will not create `index.html` if option `--use-contents` or option `--use-index` is passed to it. See: https://haskell-haddock.readthedocs.io/en/latest/invoking.html#cmdoption-use-contents

However, that is not reflected in the existing logic for `result` in `Distribution.Simple.Haddock.renderArgs`, resulting in the behaviour complained about in issue #5210. This pull request fixes that.

* [x] Patches conform to the [coding conventions](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#other-conventions). I have followed the style of the surrounding source code.
* [x] Any changes that could be relevant to users [have been recorded in the changelog](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#changelog). I don't think this fix merits a Change Log entry.
* [x] The documentation has been updated, if necessary. The feature is undocumented, so there is no documentation to update.
* [x] [Manual QA notes](https://github.com/haskell/cabal/blob/master/CONTRIBUTING.md#qa-notes) have been included. I don't think QA notes are needed, but I have included a code comment.

### CODE REVIEW

File: Cabal/src/Distribution/Simple/Haddock.hs [hunk 1]
```diff
@@ -821,14 +821,21 @@ renderArgs verbosity tmpFileOpts version comp platform args k = do
         else k (renderedArgs, result)
   where
     outputDir = (unDir $ argOutputDir args)
+    isArgContents = isJust (flagToMaybe $ argContents args)
+    isArgIndex = isJust (flagToMaybe $ argIndex args)
+    isArgGenIndex = fromFlagOrDefault False (argGenIndex args)
+    -- Haddock, when generating HTML, does not generate an index if the options
+    -- --use-contents or --use-index are passed to it. See
+    -- https://haskell-haddock.readthedocs.io/en/latest/invoking.html#cmdoption-use-contents
+    isIndexGenerated = isArgGenIndex && (not isArgContents) && (not isArgIndex)
```
Comment (geekosaur) [inline Cabal/src/Distribution/Simple/Haddock.hs:830]: Are those parentheses necessary?

**Review by geekosaur — APPROVED** [review]

**Comment by ulysses4ever** [comment]
Amy chance for a test?

Changelog would be nice too.

**Comment by mpilgrem** [comment]
As requested, I've add something to `Cabal`'s ChangeLog (I hope it was the correct file) and also to `cabal-install`'s (as it flows through to the tool, too).

I can't think of how to test for the original mistake.

**Comment by geekosaur** [comment]
There's a `changelog.d` directory into which changelog fragments go, rather than edit the `ChangeLog` directly; this lets changelogs be built for a particular version from the commits that go into that version.

**Comment by ulysses4ever** [comment]
Thanks! The explanation about changelogs is referenced in the PR template.

**Comment by mpilgrem** [comment]
@geekosaur, @ulysses4ever apologies on ChangeLog. I've now read the instructions, and tried again.

**Comment by ulysses4ever** [comment]
For testing: could you create a copy of `cabal-testsuite/PackageTests/NewHaddock/HaddockOutput/HaddockOutputCmd`, rename it, change the call to `cabal` in `cabal.test.hs` from
```
  r <- cabal' "haddock" ["--haddock-output-dir=docs", "A"] 
```
to
```
  r <- cabal' "haddock" ["--haddock-output-dir=docs", " --haddock-for-hackage" "A"] 
```
and replace the last line in the same file (`assertFindInFile...`) with `assertOutputDoesNotContain "index.html" r`?

I may be missing some details but high-level this is the way to go, I think.

To test is locally you may need to dive into https://github.com/haskell/cabal/blob/master/cabal-testsuite/README.md Most importantly, `--with-cabal` should be set correctly (e.g. to `$(cabal list-bin cabal)` after you have built `cabal`).

**Comment by mpilgrem** [comment]
@ulysses4ever, I had added a test but my attempts to test the test locally came to naught, because `Test.Cabal.Prelude.withRepo` contains a `skipIfWindows`. So, I am relying on CI to test the test.

**Comment by ulysses4ever** [comment]
Thank you for willing to give it a go, and I'm sorry it doesn't go through yet.

`withRepo` may actually be not necessary. You could try any other existing test that calls cabal against a local package instead of a dependency (`withRepo` is only necessary when a test package uses a dependency).

**Comment by ulysses4ever** [comment]
Sorry, didn't see the edits until now (better send new comments — for visibility). Yes, .out will need to be adjusted. Usually it's done by calling the test binary with --accept but if you can't run it locally, this may be a problem...

**Comment by mpilgrem** [comment]
@ulysses4ever, I finally arrived at a test that has passed the CI.

**Review by ulysses4ever — APPROVED** [review]
Terrific, thank you! Apologies for all the pain due to the test suite, but tests is the only way to assure sanity in the future..

**Pull request merged**